source/rpt/rpt_fem_reconstruction.cc

```diff
@@ -0,0 +1,1080 @@
+    }
+
+  // Loop over adjacent cells, loop over detectors and get nodal count values
+  // for each vertex of the cell
+  for (const auto &adjacent_cell : all_adjacent_cells)
+    {
+      for (unsigned int d = 0; d < n_detector; ++d)
+        {
+          for (unsigned int v = 0; v < adjacent_cell->n_vertices(); ++v)
+            {
+              auto dof_index = adjacent_cell->vertex_dof_index(v, 0);
+              count_from_all_detectors[d][v] = nodal_counts[d][dof_index];
+            }
+        }
+
+      // Solve linear system to find the location in reference coordinates
+      reference_location = assemble_matrix_and_rhs<dim>(
+        count_from_all_detectors,
+        experimental_count,
+        rpt_parameters.fem_reconstruction_param.fem_cost_function);
+
+      // 4th constraint on the location of the particle in reference coordinates
+      last_constraint_reference_location = 1 - reference_location[0] -
+                                           reference_location[1] -
+                                           reference_location[2];
+
+      // Evaluate the error of the reference position (Is it outside the
+      // reference tetrahedron ?)
+      norm_error_coordinates =
+        calculate_reference_location_error(reference_location,
+                                           last_constraint_reference_location);
+
+      // Extrapolation limit
+      if (norm_error_coordinates < tol_reference_location)
+        {
+          // Calculate cost with the selected cost function
+          calculated_cost = calculate_cost(adjacent_cell,
+                                           reference_location,
+                                           last_constraint_reference_location,
+                                           experimental_count);
+          // Check if the new cost is lower than the previously stored
+          // real_location's
+          if (calculated_cost < max_cost_function)
+            {
+              position_found         = true;
+              max_cost_function      = calculated_cost;
+              previous_position_cell = adjacent_cell;
+
+              // Evaluate the real location of the particle
+              for (unsigned int v = 0; v < adjacent_cell->n_vertices(); ++v)
+                {
+                  if (v == 0)
+                    real_location = adjacent_cell->vertex(v);
+                  else
+                    real_location +=
+                      reference_location[v - 1] *
+                      (adjacent_cell->vertex(v) - adjacent_cell->vertex(0));
+                }
+            }
+        }
+    }
+
+  if (position_found)
+    {
+      found_positions.push_back(real_location);
+      return true;
+    }
+  else
+    return false;
+}
+
+template <int dim>
+void
+RPTFEMReconstruction<dim>::read_experimental_counts(
+  std::vector<std::vector<double>> &all_experimental_counts)
+{
+  TimerOutput::Scope t(computing_timer, "read_experimental_counts");
+
+  std::ifstream in(
```

Comment:
It would be good to give a better name to this file... `input_experimental_counts_file` or something like this